### ISSUE TITLE
Wrong serial after turn a mirror authoritative and add new objects

### ISSUE THREAD

**Issue by hfaleiros** [irrdnet/irrd4#269]
**Describe the bug**

when I migrate existing data (following the documentation) everything works fine (no errors), then I set authoritative to true (also no errors) but when I add or modify an object the serial does not increment.

**To Reproduce**

Let's use TC as a sample:

1. authoritative = false, mirror OK

% whois -h ::1 '!jtc'
A17
TC:Y:29174-29174
C

2. authoritative = true

% whois -h ::1 '!jtc'
A23
TC:Y:29174-29174:29174
C

3. add or modify

% whois -h ::1 '!jtc'
A19
TC:Y:1-29174:29174
C

...and:

-g TC:3:1-LAST
%START Version: 3 TC 1-1

Note that the serial does not increment.

-g TC:3:29174-LAST        
%% ERROR: Serials 1 - 29174 do not exist

and from status page:

    Oldest serial seen: 1
    Newest serial seen: 29174
    Oldest journal serial number: 1
    Newest journal serial number: 1

**Expected behavior**

The serial should increment from 29174, wright? Not starts with 1.

**IRRd version you are running**

4.0.8.

**Comment by job** [comment]
to recap: are you switching from "mirror" to "authoritative" for a source, and after that the serial does not increment, but you *do* see the new or modified object when you query via `whois`?

**Comment by hfaleiros** [comment]
new or modified objects are there (it's work as expected too), the problem is the serial that should increment from the current serial, but it starts with 1.

**Comment by job** [comment]
if you make other modifications, does it continue to increment?

Am I correct in understanding that you essentially need a one-off serial
bump to get closer to where TC currently in reality is?

**Comment by hfaleiros** [comment]
yes, it's growing correctly but *from 1*. Should be (using my example) from 29174.

**Comment by job** [comment]
@mxsasha what is your take?

**Comment by mxsasha** [comment]
Was `keep_journal` enabled for this source, while you were running it as a mirror? I suspect not.

The `!j` query, as of #181, returns which [serial numbers have been seen](https://irrd4.readthedocs.io/en/v4.0.8/users/queries.html):
> The serial range concerns the lowest and highest serial number seen for this database. It is unrelated to the range available in the journal, if any (which is used for NRTM).

If a mirror is run for a while, and then switched to authoritative with keep_journal on, the starting point is an empty journal, which [start from serial 1](https://github.com/irrdnet/irrd4/blob/master/irrd/storage/database_handler.py#L368-L371).

This is consistent with
```
Oldest serial seen: 1
Newest serial seen: 29174
Oldest journal serial number: 1
Newest journal serial number: 1
```
The newest serial seen (as returned by `!j`) is indeed in the 29000's, but those entries are not in the journal.

**Comment by hfaleiros** [comment]
> Was `keep_journal` enabled for this source, while you were running it as a mirror? I suspect not.

no, as mirror keep_journal was true too.

> The `!j` query, as of #181, returns which [serial numbers have been seen](https://irrd4.readthedocs.io/en/v4.0.8/users/queries.html):
> 
> > The serial range concerns the lowest and highest serial number seen for this database. It is unrelated to the range available in the journal, if any (which is used for NRTM).
> 
> If a mirror is run for a while, and then switched to authoritative with keep_journal on, the starting point is an empty journal, which [start from serial 1](https://github.com/irrdnet/irrd4/blob/master/irrd/storage/database_handler.py#L368-L371).
> 
> This is consistent with
> 
> ```
> Oldest serial seen: 1
> Newest serial seen: 29174
> Oldest journal serial number: 1
> Newest journal serial number: 1
> ```
> 
> The newest serial seen (as returned by `!j`) is indeed in the 29000's, but those entries are not in the journal.

I think I found the root cause for what I reported.

I setup as mirror and as soon the db sync I'd switch it to authoritative (without any NRTM update). This way serial act as I described (starts with 1, which is indeed wrong I think).

I tested the same deploy, but now waited for some NRTM update been processed, this way any new (or modify) increments correctly when turn the mirror authoritative.

So the trick is to wait for some NRTM before switch it to authoritative!

**Comment by mxsasha** [comment]
Yes, that makes sense. No NRTM updates (yet) means no journal, so if you then start adding authoritative updates to the empty journal, they start from serial 1. This is not entirely intuitive, but an effect of how journalling/NRTM/serials work in IRR. It is fixable (for an empty journal, we could count from the "last seen" serial) but it seems like quite the edge case.

**Comment by job** [comment]
So is the conclusion is that perhaps the initial seeding of what is intended to become an *authoritative* database, should not take place over NRTM? Then perhaps this can be addressed through an update to the migration documentation.

**Comment by mxsasha** [comment]
In IRRd 4.1, we've decided to entirely decouple local journal serials from mirror source serials, to include support for RPKI and scope filtering. So IRRd will now always create it's own set of serials, rather than incrementing from the mirror source serial.
